README

          DES-SERT - an Extensible Routing-Framework for Testbeds 


Copyright
   
   Copyright Philipp Schmidt <phils@inf.fu-berlin.de>,
   Computer Systems and Telematics / Distributed, Embedded Systems (DES) group, 
   Freie Universitaet Berlin

   This document has been published under GNU Free Documentation License.   
   All rights reserved.


1.  Introduction

   DES-SERT, the DES Simple and Extensible Routing-Framework for Testbeds,
   is a framework designed to assist researchers implementing routing
   protocols for testbeds.
   
   DES-SERT enables the implementation of routing protocols on top of
   Ethernet via an underlay (Layer 2.5) in user space.
   It introduces an abstraction from OS specific issues and provides
   functionality and data structures to implement proactive, reactive,
   and hybrid routing protocols. 
   
   While generally usable in many application scenarios, it is primarily
   used in DES-Mesh (http://www.des-testbed.net/), the multi-transceiver
   wireless mesh network testbed part of the DES-Testbed. 


2.  DES-SERT Architecture

   DES-SERT introduces some concepts to implement routing protocols.
   When implementing a routing protocol with DES-SERT, you should be
   familiar with these concepts to structure and tailor your implementation.


2.1.  messages

   Every packet you send or receive on the mesh is represented as a
   DES-SERT message. From a programmers point of view, a DES-SERT message
   is just a C-structure:
   
      typedef struct dessert_msg {
          /** the layer2 header on the wire */
          struct     ether_header l2h;
          /** short name of the protocol as passed to dessert_init() */
          char       proto[DESSERT_PROTO_STRLEN];
          /** version of the app as passed to dessert_init() */
          uint8_t    ver;
          /** flags - bits 1-4 reserved for dessert, bits 5-8 for app usage */
          uint8_t    flags;
          /** ttl or hopcount field for app usage - 0xff if not used*/
          uint8_t    ttl;
          /** reserved for app usage - 0x00 if not used */
          uint8_t    u8;
          /** reserved for app usage - 0xbeef if not used */
          uint16_t   u16;
          /** header length incl. extensions */
          uint16_t   hlen;
          /** payload length */
          uint16_t   plen;
      } dessert_msg_t;
   
   Every message sent via the underlay carries this structure as a packet
   header. All data in a "dessert_msg" is stored in network byte order. 
   DES-SERT tries to care as automatically as possible of this structure.
   Nevertheless you will have to care at least about: "l2h.ether_dhost" and
   "ttl".
   
   If you need to send some data along with every packet, e.g. some kind of
   metric or cost your routing protocol uses, you should try to fit this
   data into the "u8", "u16" and the upper 4 bits of the "flags" field.
   These fields will never be touched by DES-SERT except on initialization
   via "dessert_msg_new".
   
   Because just a C-structure is not really usable as a packet, there are some
   utility functions around - please have a look around in "dessert.h" and the  
   doxygen doku. The most important ones are: "dessert_msg_new" and
   "dessert_msg_destroy", which do not simply allocate memory for a DES-SERT
   message, but for a whole packet of maximum size and initialize the 
   structures for further packet construction/processing.
   
      int dessert_msg_new(dessert_msg_t **msgout);
      
      void dessert_msg_destroy(dessert_msg_t* msg);


2.1.2  DES-SERT extensions

   A DES-SERT extension is some structure used to piggyback data on a 
   DES-SERT message. It consists of a 8-bit user supplied type field (with
   some reserved values), an 8-bit length field and user supplied data of
   arbitrary length of 253 bytes at most.
   
   It can be added to a message via "dessert_msg_addext", retrieved via
   "dessert_msg_getext" and removed via "dessert_msg_delext".
   
      int dessert_msg_addext(dessert_msg_t* msg, 
         dessert_ext_t** ext, uint8_t type, size_t len);
      
      int dessert_msg_getext(const dessert_msg_t* msg, 
         dessert_ext_t** ext, uint8_t type, int index);
      
      int dessert_msg_delext(dessert_msg_t *msg, dessert_ext_t *ext);
   
   It is recommended not to put single data fields in extensions, but 
   combine semantically related data in a struct and attach this struct
   as an extension because every extension carried introduces an 16-bit
   overhead to the packet.


2.2.  Processing pipelines

   Routing algorithms are often split up in several parts like packet 
   validation, loop-detection or routing table lookup.
   To implement these as independent and clear as possible, DES-SERT enables
   you to split up your packet processing in as many parts as you like.
   
   There are two separate processing pipelines - one for packets received 
   from the kernel via a TUN or TAP interface and one for packets received
   via an interface used on the mesh network.
   
   You can register callbacks to be added to one of these pipelines with 
   "dessert_sysrxcb_add" or "dessert_meshrxcb_add". Both take an additional
   integer argument ("priority") specifying the order the callbacks should
   be called. Higher "priority" value results in being called later
   within the pipeline.
   
      int dessert_sysrxcb_add(dessert_sysrxcb_t* c, int prio);
      
      int dessert_meshrxcb_add(dessert_meshrxcb_t* c, int prio);

   If a callback returns "DESSERT_MSG_KEEP" the packed will be processed by
   further callbacks, if it returns "DESSERT_MSG_DROP" the message will be
   dropped and no further callbacks will be called.
   
   You do not need to care about the management of the buffers for incoming
   messages - DES-SERT does this for you. Nevertheless if you need to add
   extensions or enlarge the payload of a message, you need to tell DES-SERT
   to enlarge the buffer for you if the flag "DESSERT_FLAG_SPARSE" is set on
   the message. You can do this by returning "DESSERT_MSG_NEEDNOSPARSE" from
   within a callback. The callback will be called again with a larger buffer
   and no "DESSERT_FLAG_SPARSE" flag being set.


2.2.1.  Processing buffer

   If you need to pass information along several callbacks, you can do this
   in the processing buffer passed to the the callbacks. This buffer contains
   some local processing flags ("lflags") set by the builtin callback
   "dessert_msg_ifaceflags_cb" (e.g. telling you about packet origin or if
   the packet is multicast) and 1KB of space for your callbacks to pass 
   along arbitrary data.
   
   This buffer might only be allocated after you explicitly request it - in
   this case the proc argument is NULL and you can return the value
   "DESSERT_MSG_NEEDMSGPROC" from within your callback. The callback will
   be called again with a valid processing buffer.


2.3.  Using interfaces


2.3.1. Using a TUN/TAP interface

   First you have to choose whether to use a TUN or TAP interface. TUN
   interfaces are used to exchange IPv4 / IPv6 datagrams with the kernel 
   network stack. TAP interfaces are used to exchange Ethernet frames
   with the kernel network stack. If you want to route Ethernet frames,
   you should choose a TAP interface. If you intend to implement
   a custom layer 2 to layer 3 mapping, you should use a TUN interface.
   
   Currently, you can only initialize and use a single sys (TUN/TAP) interface.
   This is done by "dessert_sysif_init". You must then set up the interface
   config in the kernel yourself e.g. by calling "ifconfig".
   
      int dessert_sysif_init(char* name, uint8_t flags);
   
   In either case, frames you receive from a TUN/TAP interface will be
   passed along the callbacks added by "dessert_sysrxcb_add" to the 
   processing pipeline. Each of them will be called with a pointer to an
   Ethernet frame. In case of a TUN interface, "ether_shost" and "ether_dhost"
   are set to "00:00:00:00:00:00", and ether_type reflects whether the packet
   received is IPv4 oder IPv6.
   
   Packets are sent to the kernel network stack with "dessert_syssend".
   In case of a TUN Interface "ether_shost" and "ether_dhost" will be
   ignored.
   
       int dessert_syssend(const struct ether_header *eth, size_t len);


2.3.2. Using a mesh interface

   Mesh interfaces are used similar to the TUN/TAP interface with two major
   differences: You can have multiple mesh interfaces and they send and
   receive DES-SERT messages instead of Ethernet frames. 
   
   You add an mesh interface using "dessert_meshif_add" and can send to it
   by calling "dessert_meshsend". If the interface parameter is NULL, the
   packet will be transmitted over every interface (good for flooding).

      int dessert_meshif_add(const char* dev, uint8_t flags);
      
      int dessert_meshsend(const dessert_msg_t* msg, 
         const dessert_meshif_t *iface);


2.4.  Logging

   You can write log messages easily with a bunch of macros provided
   by DES-SERT ("dessert_debug", "dessert_info" ,"dessert_notice",
   "dessert_warn", "dessert_warning", "dessert_err", "dessert_crit", 
   "dessert_alert" and "dessert_emerg"). Each of them can be used like
   "printf" and logs to Syslog, STDERR, file or a ringbuffer depending
   on your configuration.

   DES-SERT also ships with a custom "assert" macro which acts like
   the original macro from the standard C library and uses the logging
   mechanism described above.


2.5.  Periodics

   Periodics help you to perform maintenance or delayed tasks. A task
   consists of a callback, which will be called at the time you requested,
   and a void pointer the callback is passed. You can add these tasks by
   calling "dessert_periodic_add" or "dessert_periodic_add_delayed".


2.6.  CLI

   DES-SERT supports simple configuration and debugging of your routing
   protocol implementation by providing a Cisco like command line interface
   (cli) and a config file parser based upon it.
   This cli is realized through libcli (http://code.google.com/p/libcli/). 
   
   DES-SERT does some of the initialization of libcli. Therefore, it provides
   the main cli anchor "dessert_cli" and some anchors to add commands below
   "dessert_cli_.*". Because DES-SERT only loosely wraps libcli, you should
   make yourself familiar with libcli itself. This may be improved in further
   DES-SERT releases.
   
   You can evaluate a config file by calling "cli_file" and start a thread
   enabling a telnet-interface for DES-SERT by calling "dessert_cli_run".


2.7. Putting all together

   Now you have learned about the most important aspects of DES-SERT.
   To write your own routing protocol implementation, you need to know
   how to put all this together.
   
   You should start with a main() program parsing the command line options
   and then calling "dessert_init()". This is needed to set up DES-SERT 
   correctly. Afterwards you can register callbacks, read the config file
   and do what you like. If everything is set up, you call "dessert_run()"
   and let the event based framework do its job.
   
   If you would like to see a complete protocol implementation sample,
   have a look at the "gossiping" directory.


3. Contact & Feedback

   We love feedback - if you have patches, comments or questions,
   please contact us! Recent contact information is available on
           http://www.des-testbed.net/des-sert/


ChangeLog
20110908 - 1.07
    - Added helper functions: dessert_timevaladd, dessert_timeval2ms, dessert_ms2timeval,
      dessert_timevaladd2
    - The sysif MAC is only derived from the eth0 address, if this MAC addr is not yet
      used. This allows to run multiple daemons on the same node.
    - Added "show uptime" CLI command.
    - Added dessert_get_uptime_ms function
    - libtool version = 1.13.1

20110720 - 1.06
    - Fixed heap corruption bug
    - Replaced size_t type by uint32_t
    - Replaced some left u_int* by uint*_t
    - Replaced uint8_t by bool where semantically appropriate
    - Fixed several issues in log message prints
    - Fixed display of scheduled time for tasks with unknown function name
    - Fixed two compiler warnings due to missing function declarations in header
    - libtool version = 1.12.1

20110720 - 1.05
    - Fixed bugs in packet filter
    - Removed legacy defines that are replaced by enums
    - Replaced int return values by enum types
    - Added enum for periodics return codes
    - The filter function is now called always before the mesh callbacks are invoked
    - Added min, max macros (if they are not already defined)
    - Added token bucket implementation with 3 policies
    - Token bucket support unlimited or tail drop queueing of packets
    - "show tasks" CLI command to show the currently registered (periodic) tasks
    - Implemented data structure and functions to map function pointer to names
    - libtool version = 1.11.1


20110712 - 1.04
    - Added CLI show function for monitor configuration
    - Added packet filter support (untested)
    - Merged dessert-extra
    - Beautified code (a little bit)
    - libtool version = 1.10.1

20110621 - 1.03
    - When binding to the IPv6 socket for the CLI fails, IPv4 is tried
    - Introduced loglevel LOG_TRACE > LOG_DEBUG for even more verbose output
    - Several preprocessor macros in dessert.h are now contained in enums
    - Renamed variable from bool to boolean to enable include of stdbool.h
    - All printing of MAC addresses realized by MAC and EXPLODE_ARRAY6 macro
    - meshiface is not const -> remove const
    - libtool version = 1.9.1

20110530 - 1.02
    - configure will now fail when uthash or iw are missing
    - Introduced alternative names for some flags in dessert.h
    - libtool version = 1.8.1

20110530 - 1.01
    - Re-written monitor code
    - More monitor interface information availabe to routing protocols
    - add dessert_utils.c with dessert_parse_mac
    - Milisecond resolution for log messages
    - add alternative log output format (disabled by default)
    - General code cleanup
    - use MAC-Macro & EXPLODE_ARRAY6-Macro
    - add DL_LENGTH macro
    - remove %M printf pattern
    - Configure includes checks for more dependencies
    - Replaced mkinstalldirs with MKDIR_P
    - libtool version = 1.7.1

20110330 - 1.00
    - Merged RSSI branch / monitor interface support
    - Added uthash and utlist header check in configure.ac
    - libtool version = 1.6.1

20110305 - 0.99
    - Logs can now be flushed every x seconds when zlib is used
    - Full string is now evaluated to update the loglevel (previously only first character)
    - libtool version = 1.5.1

20110226 - 0.98
    - Added mutex for lock file writes
    - libtool version = 1.4.1

unreleased - 0.97
    - Increased maximum number of characters in a log message from 1024 to 4096
    - Current libtool version = 1.4.1

20100801 - 0.96
    - Merged Android branch; SNMP support can now be disabled
    - Fixed zlib bug
    - Removed deprecated DESSERT_LOG_DEBUG and DESSERT_LOG_NODEBUG flags
    - Unified log output format (always a newline before the function and line number is printed)
    - Current libtool version = 1.3.1

20100624 - 0.95
    - loglevel is now configurable over the CLI
    - signals can be caught and handled by registered callbacks
    - Fixed wireshark dissectors (still needs testing on Debian lenny)
    - Changed the ethertype from 0x8042 to 0x88B5 (IEEE Std 802 - Local Experimental Ethertype 1)
    - Increased DESSERT_MAXEXTDATALEN from 253 to 254 (extension can have a length of 256 bytes: data + 2 byte header)
    - DESSERT_MAXFRAMELEN is now of size ETH_FRAME_LEN and not ETHER_MAX_LEN anymore (the latter includes the 4 byte CRC)
    - The define DESSERT_MAXFRAMELEN has been replaced with the variable dessert_maxlen (initial length = ETH_FRAME_LEN)
    - Fixed bug that made it impossible to disable logging to a file
    - Log files can now be gzip'ed: either use DESSERT_LOG_GZ as option for dessert_logcfg or specify a log file with ".gz" as extension. If you use DESSERT_LOG_GZ and specify a file without ".gz" as extension, the extension will be added automatically.
    - Registered SIGTERM handler to flush and close the log file
    - Current libtool version = 1.2.1

20100606 - 0.94
    - On Linux the TAP hwaddr is now derived from eth0 (if it exists)
    - Current libtool version = 1.1.1

20100323 - 0.93
    - Updated documentation (spelling, grammar, etc)
    - Removed src/callbacks; the code is now in libdessert-extra

20100319 - 0.92
    - Updated, restructured, and extended documentation

20100214 - 0.91
    - libtool version = 1.1.0
    - Added some code from several daemons in src/callbacks dir for general usage
    - Moved some code from src/libdessert to src/callbacks to slim down library core
    - Moved prev pointer in dessert_meshif_t to end of structure; dessert_meshif_t and dessert_sysif_t are
      now equal again in the first few fields
    - Added type parameter to dessert_msg_trace_initiate and dessert_msg_trace_dump to use these function with new
      bidirectional tracing feature
    - Added callback that decrements the IPv4 TTL or IPv6 Hop-Limit field
    - Removed pid file parameter from dessert_init; use dessert_pid to write pid file manually
    - Removed DESSERT_OPT_PID and DESSERT_OPT_NOPID flags because they were never used

dissectors/dessert.lua
-- ----------------------------------------------------------------------------
-- Copyright 2009, David Gutzmann, Freie Universitaet Berlin (FUB).
-- Copyright 2010, Bastian Blywis, Freie Universitaet Berlin (FUB).
-- All rights reserved.

--These sources were originally developed by David Gutzmann,
--rewritten and extended by Bastian Blywis
--at Freie Universitaet Berlin (http://www.fu-berlin.de/),
--Computer Systems and Telematics / Distributed, embedded Systems (DES) group
--(http://cst.mi.fu-berlin.de, http://www.des-testbed.net)
-- ----------------------------------------------------------------------------
--This program is free software: you can redistribute it and/or modify it under
--the terms of the GNU General Public License as published by the Free Software
--Foundation, either version 3 of the License, or (at your option) any later
--version.

--This program is distributed in the hope that it will be useful, but WITHOUT
--ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
--FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

--You should have received a copy of the GNU General Public License along with
--this program. If not, see http://www.gnu.org/licenses/ .
-- ----------------------------------------------------------------------------
--For further information and questions please use the web site
--       http://www.des-testbed.net
-- ----------------------------------------------------------------------------

local extension_types = { [0x01] = "DESSERT_EXT_ETH",
                          [0x02] = "DESSERT_EXT_TRACE_REQ",
                          [0x03] = "DESSERT_EXT_TRACE_RPL",
                          [0x04] = "DESSERT_EXT_PING",
                          [0x05] = "DESSERT_EXT_PONG",
                          [0x40] = "DESSERT_EXT_USER"
                         }

dessert_dissector_table = DissectorTable.new("desserttable")

_G.dessert_register_ext_dissector = function(ext_type, ext_name, dissector)
    print("Info: loading extension dissector: "..tostring(ext_name))
    extension_types[ext_type] = ext_name
    dessert_dissector_table:add(ext_type, dissector)
end

dofile("dessert-ext-eth.lua")
dofile("dessert-ext-ping.lua")
dofile("dessert-ext-trace.lua")

-- Create a new dissector
DESSERT = Proto ("dessert", "DES-SERT")

-- Create the protocol fields
local f = DESSERT.fields

f.proto   = ProtoField.string ("dessert.proto"  , "Protocol name")
f.version = ProtoField.uint8  ("dessert.version", "Protocol version")

f.flags   = ProtoField.uint8  ("dessert.flags"  , "Flags" , base.HEX, nil, 0xFF)
f.flags_1 = ProtoField.uint8  ("dessert.flags_1", "Flag 1", base.HEX, nil, 0x01)
f.flags_2 = ProtoField.uint8  ("dessert.flags_2", "Flag 2", base.HEX, nil, 0x02)
f.flags_3 = ProtoField.uint8  ("dessert.flags_3", "Flag 3", base.HEX, nil, 0x04)
f.flags_4 = ProtoField.uint8  ("dessert.flags_4", "Flag 4", base.HEX, nil, 0x08)
f.flags_5 = ProtoField.uint8  ("dessert.sparse" , "DESSERT_FLAG_SPARSE", base.HEX, nil, 0x10,"if not set buffer len is assumed as DESSERT_MAXFRAMELEN + DESSERT_MSGPROCLEN ")
f.flags_6 = ProtoField.uint8  ("dessert.flags_6", "Flag 6", base.HEX, nil, 0x20)
f.flags_7 = ProtoField.uint8  ("dessert.flags_7", "Flag 7", base.HEX, nil, 0x40)
f.flags_8 = ProtoField.uint8  ("dessert.flags_8", "Flag 8", base.HEX, nil, 0x80)

f.u32     = ProtoField.uint32 ("dessert.u32", "u32", base.HEX ,nil)
f.ttl     = ProtoField.uint8  ("dessert.ttl", "ttl", base.DEC ,nil)
f.u8      = ProtoField.uint8  ("dessert.u8", "u8", base.HEX ,nil)
f.u16     = ProtoField.uint16 ("dessert.u16", "u16", base.HEX, nil)

f.hlen    = ProtoField.uint16 ("dessert.hlen", "Header length hlen (incl. layer 2.5)", base.DEC, nil,nil,"header length incl. extensions")
f.plen    = ProtoField.uint16 ("dessert.plen", "Payload length plen", base.DEC, nil)
f.exts    = ProtoField.string ("dessert.exts", "Extensions")

EXTHDR = Proto("dessert_ext", "DESSERT_EXT");
local e = EXTHDR.fields
e.exttype = ProtoField.uint8  ("dessert.ext.type", "Extension type", base.HEX )
e.extlen  = ProtoField.uint8  ("dessert.ext.len", "Extension length (incl. extension header)", base.DEC ,nil,nil, "length of the extension in bytes, including the 2 bytes of the extension header itself")

-- The dissector function
function DESSERT.dissector(buffer, pinfo, tree)
    pinfo.cols.protocol = "DES-SERT"

    local subtree = tree:add(DESSERT, buffer(),"DES-SERT Protocol Data")
    local offset = 0

    local proto = buffer (offset, 4)
    subtree:add(f.proto, proto)
    offset = offset + 4

    local version= buffer (offset, 1)
    subtree:add(f.version, version)
    offset = offset + 1

    local flags = buffer (offset, 1)
    local flags_field = subtree:add(f.flags, flags)
    flags_field:add(f.flags_1, flags)
    flags_field:add(f.flags_2, flags)
    flags_field:add(f.flags_3, flags)
    flags_field:add(f.flags_4, flags)
    flags_field:add(f.flags_5, flags)
    flags_field:add(f.flags_6, flags)
    flags_field:add(f.flags_7, flags)
    flags_field:add(f.flags_8, flags)
    offset = offset + 1

    local u32 = buffer (offset, 4)
    local ttl = buffer (offset, 1)
    local u8  = buffer (offset+1, 1)
    local u16 = buffer (offset+2, 2)
    local u32_field = subtree:add(f.u32, u32)
    u32_field:add(f.ttl, ttl)
    u32_field:add(f.u8 , u8)
    u32_field:add(f.u16, u16)
    offset = offset + 4

    local hlen = buffer(offset,2)
    subtree:add(f.hlen, hlen)
    offset = offset + 2

    -- because wireshark already dissected the layer 2.5 header at this point
    -- the *real* header length is hlen-14
    local real_hlen = hlen:uint() - 14

    local plen = buffer(offset,2)
    subtree:add(f.plen, plen)
    offset = offset + 2

    extensions = subtree:add(f.exts)

    local extension, exttype, extlen, extdata_real_length, extdata, exttreeitem, dissector
    local ext_count = 0

    while offset < real_hlen do
      ext_count = ext_count +1
      extension = extensions:add(EXTHDR)
      extension:set_generated()

      exttype = buffer(offset, 1)
      extension:add(e.exttype, exttype)
      offset = offset + 1

      extlen = buffer(offset, 1)
      extension:add(e.extlen, extlen)
      offset = offset + 1

      -- extlen includes the 2byte extension header !
      extdata_real_length = extlen:uint() - 2
      extdata = buffer(offset, extdata_real_length)
      local dissector = dessert_dissector_table:get_dissector(exttype:uint())
      if dissector ~= nil then
          dissector:call(extdata:tvb(), pinfo, extension)
		  length_dissected = _G.g_offset
		  ethertype = _G.g_ethertype
          if length_dissected ~= extdata_real_length then
            print("\t\t\tWarning: Sub-Dissector did not consume all bytes!")
          end
      else
          print("\t\t\tWarning: No extension_dissector for ext_type: "..tostring(extension_types[exttype:uint()]))
      end
      offset = offset + extdata_real_length
      -- print("\t\toffset="..tostring(offset)..", hlen="..tostring(real_hlen))
    end

    -- print("\tno more extensions, offset="..tostring(offset))
    -- dissect paylod based on ext_eth ethertype if any
    ethertype = _G.g_ethertype
    if ethertype:uint() ~= 0 then
        local dissector = ethertype_table:get_dissector(ethertype:uint())
        if dissector ~= nil then
            local payload = buffer(offset, plen:uint())
            dissector:call(payload:tvb(), pinfo, tree)
        else
          print("Warning: Payload found but no matching dissector")
        end
    end
    return offset
end

-- load the ethertype table
ethertype_table = DissectorTable.get("ethertype")

-- register DES-SERT protocol to handle ethertype 0x88B5
ethertype_table:add(0x88B5, DESSERT)

print("DES-SERT dissector loaded")
